### ISSUE TITLE
f32.round/f64.round probably does not round correctly

### ISSUE THREAD

**Issue by Athas** [diku-dk/futhark#377]
It is unsurprisingly fairly complicated to round floating-point numbers correctly: https://www.cockroachlabs.com/blog/rounding-implementations-in-go/

We should ensure that we can pass the cases in that blog post.

**Comment by melsman** [comment]
... the present implementation follows the specification given by the
Standard ML Basis library...

>

**Comment by Athas** [comment]
It looks like it is wrong for infinity and NaN values, at least.

**Comment by Athas** [comment]
`round` works for NaN, but is wrong for infinities, and for e.g. `0.5000000000000001` (epsilon above `0.5`).

Both `floor` and `ceil` behave oddly when given infinities (returning `0` or `1`).

**Comment by Athas** [comment]
Here's the beginnings of more nasty testing:

```
-- ==
-- entry: test_f64_round_weird
-- input {  0.0 0.0 } output { false true false }
-- input {  1.0 0.0 } output { true false true }
-- input { -1.0 0.0 } output { false false true }

entry test_f64_round_weird (x: f64) (y: f64) =
  let z = f64.round (x / y)
  in (z > 0f64, f64.isnan z, f64.isinf z)
```

We don't pass this yet.

**Comment by Athas** [comment]
@melsman will you have time to look at this at some point?

**Comment by Athas** [comment]
It is kind of embarassing that we still cannot round numbers properly in Futhark.  Although I guess we're in good company, since both Haskell and Standard ML also round numbers uncorrectly.  I don't think we'll fix this in time for 0.4.0, as I originally hoped, but I hope we can at least fix it before the issue grows a year old.

**Comment by Athas** [comment]
Darn, we missed it by a few days - the one-year anniversary of this bug was a few days ago.  But now I have a solution.  Just writing up the final tests for the single-precision case.
